lib/apiservers/engine/backends/container_proxy.go

```diff
@@ -776,72 +773,85 @@ func (c *ContainerProxy) AttachStreams(ctx context.Context, vc *viccontainer.Vic
 		}
 	}
 
+	var inCtx context.Context
+	var inCancel, cancel func()
 	if ca.UseStdin {
 		wg.Add(1)
+		inCtx, inCancel = context.WithCancel(ctx)
+		defer inCancel()
+	}
+
+	if ca.UseStdout {
+		wg.Add(1)
+	}
+
+	if ca.UseStderr {
+		wg.Add(1)
+	}
+
+	if ca.UseStderr || ca.UseStdout {
+		ctx, cancel = context.WithCancel(ctx)
+		defer cancel()
+	}
+
+	if ca.UseStdin {
 		go func() {
 			defer wg.Done()
-			err := copyStdIn(ctx, plClient, vc, clStdin, keys)
+			err := copyStdIn(inCtx, plClient, vc, clStdin, keys)
 			if err != nil {
-				log.Errorf("container attach: stdin (%s): %s", vc.ContainerID, err.Error())
+				log.Errorf("container attach: stdin (%s): %s", vc.ContainerID, err)
 			} else {
-				log.Infof("container attach: stdin (%s) done: %s", vc.ContainerID)
+				log.Infof("container attach: stdin (%s) done", vc.ContainerID)
 			}
 
-			// no need to take action if we are canceled
-			// as that means error happened somewhere else
-			if ctx.Err() == context.Canceled {
-				log.Infof("returning from stdin as context canceled somewhere else")
-				return
+			if !vc.Config.StdinOnce || vc.Config.Tty {
+				// close stdout/stderr
+				cancel()
 			}
-			cancel()
 
-			errors <- err
+			if err != context.Canceled && err != io.EOF {
+				errors <- err
+			}
 		}()
 	}
 
 	if ca.UseStdout {
-		wg.Add(1)
 		go func() {
 			defer wg.Done()
 			err := copyStdOut(ctx, plClient, attachAttemptTimeout, vc, clStdout)
 			if err != nil {
-				log.Errorf("container attach: stdout (%s): %s", vc.ContainerID, err.Error())
+				log.Errorf("container attach: stdout (%s): %s", vc.ContainerID, err)
 			} else {
-				log.Infof("container attach: stdout (%s) done: %s", vc.ContainerID)
+				log.Infof("container attach: stdout (%s) done", vc.ContainerID)
 			}
 
-			// no need to take action if we are canceled
-			// as that means error happened somewhere else
-			if ctx.Err() == context.Canceled {
-				log.Infof("returning from stdin as context canceled somewhere else")
-				return
+			if inCancel != nil {
+				inCancel()
 			}
-			cancel()
 
-			errors <- err
+			if err != context.Canceled && err != io.EOF {
+				errors <- err
+			}
 		}()
 	}
 
 	if ca.UseStderr {
-		wg.Add(1)
 		go func() {
 			defer wg.Done()
 			err := copyStdErr(ctx, plClient, vc, clStderr)
 			if err != nil {
-				log.Errorf("container attach: stderr (%s): %s", vc.ContainerID, err.Error())
+				log.Errorf("container attach: stderr (%s): %s", vc.ContainerID, err)
 			} else {
-				log.Infof("container attach: stderr (%s) done: %s", vc.ContainerID)
+				log.Infof("container attach: stderr (%s) done", vc.ContainerID)
 			}
 
-			// no need to take action if we are canceled
-			// as that means error happened somewhere else
-			if ctx.Err() == context.Canceled {
-				log.Infof("returning from stdin as context canceled somewhere else")
-				return
+			if inCancel != nil {
+				inCancel()
```

Comment:
We may actually not need these and just use a common context for all three stream copy loops. That way if one gets cancelled, all of them get cancelled; I can only see that happening if the client connection drops or times out.